### ISSUE TITLE
Support for standard `Idempotency-Key` header

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#2998]
## Problem

Currently idempotent requests (GET, HEAD) are retryable but not non-idempotent requests like POST or PATCH.

## Solution

Add support for specifying the standard `Idempotency-Key` header. The RFC is currently in [draft](https://greenbytes.de/tech/webdav/draft-ietf-httpapi-idempotency-key-header-latest.html#uniqueness-of-idempotency-key) but some services already support it:

- https://stripe.com/docs/api/idempotent_requests
- https://developers.squarespace.com/commerce-apis/idempotency-key

The main idea:

```bash
# 1st request
POST /projects
Authorization: Bearer $JWT
Prefer: return=representation
Idempotency-Key: 8e03978e-40d5-43e8-bc93-6894a57f9324

{"name": "project-1"}

# generated query
INSERT INTO "test"."projects"("id", "name")
SELECT "pgrst_body"."id", "pgrst_body"."name" FROM (SELECT '{"id": 346, "name": "project-1"}'::json AS json_data) pgrst_payload,
LATERAL (SELECT CASE WHEN json_typeof(pgrst_payload.json_data) = 'array' THEN pgrst_payload.json_data ELSE json_build_array(pgrst_payload.json_data) END AS val) pgrst_uniform_json,
LATERAL (SELECT "id", "name" FROM json_to_recordset(pgrst_uniform_json.val) AS _("id" integer, "name" text) ) pgrst_body
RETURNING "test"."projects".*;

HTTP/1.1 201 Created
{"id": 345, "name": "project-1"}

# 2nd request

POST /projects
Authorization: Bearer $JWT
Prefer: return=representation
Idempotency-Key: 8e03978e-40d5-43e8-bc93-6894a57f9324

HTTP/1.1 201 Created
{"id": 345, "name": "project-1"}

# generated query, no INSERT this time
SELECT "pgrst_body"."id", "pgrst_body"."name" FROM (SELECT '{"id": 346, "name": "project-1"}'::json AS json_data) pgrst_payload,
LATERAL (SELECT CASE WHEN json_typeof(pgrst_payload.json_data) = 'array' THEN pgrst_payload.json_data ELSE json_build_array(pgrst_payload.json_data) END AS val) pgrst_uniform_json,
LATERAL (SELECT "id", "name" FROM json_to_recordset(pgrst_uniform_json.val) AS _("id" integer, "name" text) ) pgrst_body
```

## Implementation Notes

- The idempotency key must be an UUID. The [rfc](https://greenbytes.de/tech/webdav/draft-ietf-httpapi-idempotency-key-header-latest.html#uniqueness-of-idempotency-key) recommends UUIDs. Some services support long strings additionally to uuid. e.g. Stripe supports 255 chars string ([ref](https://stripe.com/docs/api/idempotent_requests)) and squarespace supports 64 char string ([ref](https://developers.squarespace.com/commerce-apis/idempotency-key)). Let's avoid the inconsistency and only support UUID. 
- The [idempotency fingerprint](https://greenbytes.de/tech/webdav/draft-ietf-httpapi-idempotency-key-header-latest.html#idempotency-fingerprint) will be the JWT for a better uniqueness guarantee. This makes sense since we already have [JWT caching](https://github.com/PostgREST/postgrest/pull/2928).
- The [idempotency expiry](https://greenbytes.de/tech/webdav/draft-ietf-httpapi-idempotency-key-header-latest.html#rfc.section.2.3) must be configurable. On https://github.com/PostgREST/postgrest/pull/2928/ we introduced the `jwt-cache-max-lifetime`, perhaps we can rename the config and reuse it somehow? Coincidentally, [we thought of a 24 hour lifetime](https://github.com/PostgREST/postgrest/pull/2928#discussion_r1333420243), and Stripe also has a 24 hour limit for the idempotency key caching ([ref](https://stripe.com/docs/api/idempotent_requests)).

**Comment by steve-chavez** [comment]
> The [idempotency fingerprint](https://greenbytes.de/tech/webdav/draft-ietf-httpapi-idempotency-key-header-latest.html#idempotency-fingerprint) will be the JWT for a better uniqueness guarantee.

The above looks wrong, the [RFC](https://greenbytes.de/tech/webdav/draft-ietf-httpapi-idempotency-key-header-latest.html#idempotency-fingerprint) recommends:

```
An idempotency fingerprint generation algorithm may use one of the following or similar approaches to create a fingerprint.

Checksum of the entire request payload.
Checksum of selected element(s) in the request payload.
Field value match for each field in the request payload.
Field value match for selected element(s) in the request payload.
Request digest/signature.
```

I think we need the checksum for this to work. Also considering the request path.

**Comment by steve-chavez** [comment]
> This will only work reliably, if:

> the table has a primary key
> you cache the idempotency-key + primary key combo

That sounds good, that way we don't have to parse the request body. We could just get the id with `RETURNING`.

Seems this should work for batch inserts too? We'd do a checksum of all the pks in the payload + idempotency-key plus cache it.

This should guarantee enough uniqueness I believe, so we don't need to use the JWT.

**Comment by taimoorzaeem** [comment]
Should we have an upper bound for how many keys we store in memory and what eviction policy to use?

**Comment by steve-chavez** [comment]
@taimoorzaeem Good questions, we'll definitely need those.

Unlike the JWT cache, which is for performance, this "Idempotency cache" is for correctness, so I don't think we can use an LRU cache. Instead we would have to fail once we hit an upper bound. 

I think we should try to reuse the logic on https://github.com/PostgREST/postgrest/pull/3802, to calculate the memory of the "Idempotency cache" (we'll need a new metric too) and then purge it given a lifetime - we would new a new config for this.

**Comment by mkleczek** [comment]
> Furthermore, an in-memory cache for idempotency keys doesn’t scale horizontally.

TBH I can't see how in-memory cache could work in this case at all in multi-instance PostgREST deployments (eg. k8s). Idempotency keys must be shared by all instances and the only place we have for shared data is the database itself.

I would say that handling of concurrent access and modifications to data and its semantic integrity is the sole responsibility of DBMS and, TBH, I am not convinced this RFC and `Idempotency-Key` header is relevant to PostgREST at all. It already supports UPSERTs.

**Comment by steve-chavez** [comment]
>  I am not convinced this RFC and Idempotency-Key header is relevant to PostgREST at all. It already supports UPSERTs.

Yes but not all POSTs are upserts and we also have PATCH so the problem of retries remains in those cases.

> TBH I can't see how in-memory cache could work in this case at all in multi-instance PostgREST deployments (eg. k8s). Idempotency keys must be shared by all instances and the only place we have for shared data is the database itself.

Hm, so maybe this should be done at the proxy/nginx level instead? Or should we have first class support for multi-database so it's not necessary to have multi-instance?

**Comment by mkleczek** [comment]
> > I am not convinced this RFC and Idempotency-Key header is relevant to PostgREST at all. It already supports UPSERTs.
> 
> Yes but not all POSTs are upserts and we also have PATCH so the problem of retries remains in those cases.

We do have Prefer: resolution

> 
> > TBH I can't see how in-memory cache could work in this case at all in multi-instance PostgREST deployments (eg. k8s). Idempotency keys must be shared by all instances and the only place we have for shared data is the database itself.
> 
> Hm, so maybe this should be done at the proxy/nginx level instead? Or should we have first class support for multi-database so it's not necessary to have multi-instance?

I’d say statelessness and multi-instance deployments are must-haves for availability and scalability. Stateful PostgREST is a no-go for k8s deployments.

Wherever you implement it, you need a shared store for the keys. What’s more: updates to this store have to be in the same tx as database updates.

It looks to me this is the kind of situation that whatever the reasons you want to move processing outside the database, don’t.
